### DISCUSSION TITLE
Run commands on imported music files

### DISCUSSION THREAD

**Discussion by df-a** [beetbox/beets#4265] [Q&A] [answered]
I'm trying to run commands on the files that are imported. Specifically say I import an album which has one or more CDs, I do the following manually:

- cd into the directory where the files have been copied to, say `/Music/Albums/My Great Album/`
- run `app1 -x *`, which will generate a single file app1.ext for `*`
- while still in the dir, run `app2 > file.ext`
- I'd also want to run variations, such as run an app on every single item with results as `item.result`

These actions have to take place inside the directory where the files are. I'm trying to keep the examples vague as to make the solutions as universal as possible. What would be the correct way to apply this?

I have tried with the `hook` command, and the commands do execute, but I see no output in the music folder.
Also, when running beet -vv, I see that the `album.path` variable for the cd would show as `/Music/Albums/My Great Album/` rather than the actual cd which would be terminal friendly, in this case `/Music/Albums/My\ Great\ Album/`

Example: `cd '{album.path}' && app1 -x * && app2 file.wma > file.ext` yields no output. How do I automatically run apps/scripts on the imported music during/after import? I need to be able to target the individual files and run scripts from inside the `{album.path}` targeting each individual `{item}`per album.

**Comment by jackwilsdon** [comment, marked as answer]
`hook` doesn't run commands in a shell, so if your configuration looks like this:
```
hook:
  hooks:
    - event: album_imported
      command: cd '{album.path}' && app1 -x * && app2 file.wma > file.ext
```

The command that is running is `cd` with the arguments `'{album.path}' && app1 -x * && app2 file.wma > file.ext`. To run multiple commands sequentially you'll want to run them in a shell.
```
command: sh -c 'cd "$0" && app1 -x * && app2 file.wma > file.ext' {album.path}
```

In terms of "safe" filenames - `hook` passes the arguments straight into the command without going through a shell, so they don't need any escaping. If you want to run in a shell, you'll need to pass arguments in like above.

I don't think there's an easy way to run for each item during an import at the moment.
